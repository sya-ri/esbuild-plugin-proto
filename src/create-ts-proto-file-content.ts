import protobuf from 'protobufjs';
import type { Options } from '@grpc/proto-loader';

export function createTSProtoFileContent(
  protoFileContent: string,
  options: Options
): string {
  const parsed = protobuf.parse(protoFileContent);

  return [
    `import { fromJSON } from '@grpc/proto-loader';`,
    `export default fromJSON(${JSON.stringify(
      parsed.root.toJSON(),
      undefined,
      2
    )}, ${JSON.stringify(options, undefined, 2)});`,
  ].join('\n');
}
