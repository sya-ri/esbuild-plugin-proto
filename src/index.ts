import { Plugin } from 'esbuild';
import { promises as fs } from 'fs';
import path from 'path';
import type { Options } from '@grpc/proto-loader';
import { createTSProtoFileContent } from './create-ts-proto-file-content';

export type ProtoPluginOptions = Options;

export function protoPlugin(options: ProtoPluginOptions): Plugin {
  return {
    name: 'proto',
    setup(build) {
      build.onLoad({ filter: /\.proto$/ }, async (args) => {
        const fileContent = await fs.readFile(args.path, 'utf8');

        const contents = createTSProtoFileContent(fileContent, options);

        return {
          contents,
          loader: 'ts',
          resolveDir: path.dirname(args.path),
        };
      });
    },
  };
}
