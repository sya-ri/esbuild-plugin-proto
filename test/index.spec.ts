import { build, BuildResult } from 'esbuild';
import { promises as fs } from 'fs';
import path from 'path';

import { protoPlugin, ProtoPluginOptions } from '../src';

describe('test', () => {
  it('should generate result given empty proto file', async () => {
    await runTestCase('empty-proto');
  });

  it('should generate result given message proto file', async () => {
    await runTestCase('message-proto');
  });

  it('should generate result given multiple message proto files', async () => {
    await runTestCase('multi-protos');
  });

  it('should generate result given proto files with import', async () => {
    await runTestCase('import-proto');
  });
});

async function runTestCase(caseName: string) {
  const outFilename = path.resolve(
    __dirname,
    `../temp/${caseName}/${new Date().toISOString()}.js`
  );

  const buldResult = await runBuild(
    path.resolve(__dirname, `./cases/${caseName}/index.js`),
    outFilename,
    {}
  );

  expect(buldResult).toBeDefined();
  expect(await fs.readFile(outFilename, 'utf8')).toMatchSnapshot();
}

function runBuild(
  entryPoint: string,
  outfile: string,
  options: ProtoPluginOptions
): Promise<BuildResult> {
  return build({
    entryPoints: [entryPoint],
    bundle: true,
    platform: 'node',
    external: ['@grpc/proto-loader'],
    outfile,
    plugins: [protoPlugin(options)],
  });
}
