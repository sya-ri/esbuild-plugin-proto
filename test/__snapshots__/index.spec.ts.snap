// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should generate result given empty proto file 1`] = `
"// test/cases/empty-proto/file.proto
var import_proto_loader = require(\\"@grpc/proto-loader\\");
var file_default = (0, import_proto_loader.fromJSON)({}, {});

// test/cases/empty-proto/index.ts
console.log(file_default);
"
`;

exports[`test should generate result given message proto file 1`] = `
"// test/cases/message-proto/file.proto
var import_proto_loader = require(\\"@grpc/proto-loader\\");
var file_default = (0, import_proto_loader.fromJSON)({
  \\"nested\\": {
    \\"Test0Response\\": {
      \\"fields\\": {
        \\"details\\": {
          \\"type\\": \\"Test0Details\\",
          \\"id\\": 1
        }
      }
    },
    \\"Test0Details\\": {
      \\"fields\\": {
        \\"id\\": {
          \\"type\\": \\"int32\\",
          \\"id\\": 1
        },
        \\"field0\\": {
          \\"type\\": \\"string\\",
          \\"id\\": 2
        }
      }
    }
  }
}, {});

// test/cases/message-proto/index.ts
console.log(file_default);
"
`;

exports[`test should generate result given multiple message proto files 1`] = `
"// test/cases/multi-protos/file0.proto
var import_proto_loader = require(\\"@grpc/proto-loader\\");
var file0_default = (0, import_proto_loader.fromJSON)({
  \\"nested\\": {
    \\"Test0Response\\": {
      \\"fields\\": {
        \\"details\\": {
          \\"type\\": \\"Test0Details\\",
          \\"id\\": 1
        }
      }
    },
    \\"Test0Details\\": {
      \\"fields\\": {
        \\"id\\": {
          \\"type\\": \\"int32\\",
          \\"id\\": 1
        },
        \\"field0\\": {
          \\"type\\": \\"string\\",
          \\"id\\": 2
        }
      }
    }
  }
}, {});

// test/cases/multi-protos/file1.proto
var import_proto_loader2 = require(\\"@grpc/proto-loader\\");
var file1_default = (0, import_proto_loader2.fromJSON)({
  \\"nested\\": {
    \\"Test1Response\\": {
      \\"fields\\": {
        \\"details\\": {
          \\"type\\": \\"Test1Details\\",
          \\"id\\": 1
        }
      }
    },
    \\"Test1Details\\": {
      \\"fields\\": {
        \\"id\\": {
          \\"type\\": \\"int32\\",
          \\"id\\": 1
        },
        \\"field1\\": {
          \\"type\\": \\"string\\",
          \\"id\\": 2
        }
      }
    }
  }
}, {});

// test/cases/multi-protos/index.ts
console.log(file0_default, file1_default);
"
`;

exports[`test should generate result given proto files with import 1`] = `
"// test/cases/import-proto/response.proto
var import_proto_loader = require(\\"@grpc/proto-loader\\");
var response_default = (0, import_proto_loader.fromJSON)({
  \\"nested\\": {
    \\"Test0Response\\": {
      \\"fields\\": {
        \\"details\\": {
          \\"type\\": \\"Test0Details\\",
          \\"id\\": 1
        }
      }
    }
  }
}, {});

// test/cases/import-proto/index.ts
console.log(response_default);
"
`;
